### PR TITLE
Fixed #24754 -- Added support for per-app permissions (not tied to a model)

### PR DESCRIPTION
**Pull request by funkybob**
- Allow Permission.content_type to be Null
- Add settings.GLOBAL_PERMS as a list of permission details
- Update a few places to handle ContentType being None.

### CODE REVIEW

File: docs/ref/settings.txt [hunk 1]
```diff
@@ -2750,6 +2750,17 @@ The model to use to represent a User. See :ref:`auth-custom-user`.
     the app that it lives in.
     See :ref:`auth-custom-user` for more details.
 
+.. setting:: GLOBAL_PERMS
+
+``GLOBAL_PERMS``
+----------------
+
+Default: ``[]``
```
Comment (felixxm) [inline docs/ref/settings.txt]: Maybe we should add _"..(Empty list)"_ like in other settings.

File: django/contrib/auth/management/__init__.py [hunk 2]
```diff
@@ -61,6 +62,9 @@ def create_permissions(app_config, verbosity=2, interactive=True, using=DEFAULT_
         ctypes.add(ctype)
         for perm in _get_all_permissions(klass._meta):
             searched_perms.append((ctype, perm))
+    # Include GLOBAL_PERMS
+    for code, name in getattr(settings, 'GLOBAL_PERMS', []):
+        ctypes.add(None, (code, name))
```
[the remark this answers is not in the document]
Comment (funkybob) [reply django/contrib/auth/management/__init__.py]: Yeah, that looks right.

File: docs/ref/settings.txt [hunk 3]
```diff
@@ -2750,6 +2750,17 @@ The model to use to represent a User. See :ref:`auth-custom-user`.
...
+
+A list of permissions to be created which are not related to any Model.
+
+It should be a tuple of 2-tuples, just like :func:`~django.db.models.Options.permissions`.
```
[the remark this answers is not in the document]
Comment (funkybob) [reply docs/ref/settings.txt]: In the permissions docs it refers to it this way, so I just repeated that.

File: tests/auth_tests/test_auth_backends.py [hunk 4]
```diff
@@ -108,6 +108,18 @@ def test_custom_perms(self):
         self.assertIs(user.has_perm('test'), False)
         self.assertIs(user.has_perms(['auth.test2', 'auth.test3']), False)
 
+    def test_global_perms(self):
```
[the remark this answers is not in the document]
Comment (funkybob) [reply tests/auth_tests/test_auth_backends.py]: It tests that a Permission created with content_type=None comes though in the get_*_permissions calls correctly, and shows as we expect -- 'None.{code}'

File: docs/ref/settings.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (funkybob) [reply docs/ref/settings.txt]: As for func, I couldn't find the proper way to reference it, and this worked... what should it be?

File: docs/releases/2.1.txt [hunk 5]
```diff
@@ -48,6 +48,11 @@ Minor features
 * :djadmin:`createsuperuser` now gives a prompt to allow bypassing the
   :setting:`AUTH_PASSWORD_VALIDATORS` checks.
 
+* The :attr:`.Permission.content_type` field is now nullable.
+
+* New setting :setting:`GLOBAL_PERMS` allows you to specify permissions not
```
[the remark this answers is not in the document]
Comment (funkybob) [reply docs/releases/2.1.txt]: Done

File: docs/releases/2.1.txt [hunk 6]
```diff
@@ -48,6 +48,11 @@ Minor features
 * :djadmin:`createsuperuser` now gives a prompt to allow bypassing the
   :setting:`AUTH_PASSWORD_VALIDATORS` checks.
 
+* The :attr:`.Permission.content_type` field is now nullable.
```
[the remark this answers is not in the document]
Comment (funkybob) [reply docs/releases/2.1.txt]: Yeah, will flesh this out once the ML discussion finishes... I like Markus' idea of adding an app_label field.

File: tests/auth_tests/settings.py [hunk 7]
```diff
@@ -16,3 +16,7 @@
         ],
     },
 }]
+
+GLOBAL_PERMS = [
+    ('can_see_view', 'Can see special View'),
```
[the remark this answers is not in the document]
Comment (funkybob) [reply tests/auth_tests/settings.py]: Fixed

**Comment by funkybob** [comment]
Yeah, I think I might refactor the create_permissions code so I can reuse half of it... and add something like "update_global_permissions", have them as two separate actions.

Totally agree I had not got far enough to consider "fixing" stale permission removal.  Mostly wanted a PoC to help start a discussion.

**Comment by funkybob** [comment]
After more discussion on the ML, I'm taking this in a different direction.  By adding an app_label field we can create per-app permissions.

With a bit more work, we can make these behave in all ways as per-model permissions.

Currently, new permissions are defined on the AppConfig, and created with django.contrib.auth.management.create_app_permissions()

**Comment by ngnpope** [comment]
>  Mostly wanted a PoC to help start a discussion.

Absolutely. Was just brain dumping as I've had thoughts on this before. We've been using something akin to the `GLOBAL_PERMS` approach with custom scripts to manage permission creation (and diffing permissions applied to users/groups between environments which is outside of the scope of this).

I like the look of this approach of per-app permissions, and will have to catch up on the mailing list.

One issue that needs to be dealt with is conflicts between per-app and per-model permissions. Per-model permission names are of the form `can_<action>_<model>`, where action is currently `add`, `change` or `delete`. (Note that `view` is planned in PR #6734.) Either a system check should be added to prevent this for existing models, or the more heavy-handed approach of disallowing permissions starting with `can_<action>_` could be used.

**Comment by funkybob** [comment]
Yeah, on the ML I raised the possibility [still viable] of using a different syntax, but then there's the question of how that works in templates.

At first blush, I'd say yes, have a warning, but basically it's up to you to manage name conflicts within you own app...

File: django/contrib/auth/migrations/0010_alter_permission_content_type_null.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    operations = [
+        migrations.AlterField(
+            model_name='permission',
+            name='content_type',
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE,
+                                    to='contenttypes.ContentType', verbose_name='content type'),
+        ),
+        migrations.AddField(
+            model_name='permission',
+            name='app_labem',
```
Comment (Stranger6667) [inline django/contrib/auth/migrations/0010_alter_permission_content_type_null.py]: ``app_labem`` -> ``app_label``

**Comment by carltongibson** [comment]
Hey @funkybob — are you still interested in working on this? If so could I ask you to update (at your leisure) and re-open a fresh one? Thanks!

**Pull request closed without merging**